mono/metadata/threads.c

```diff
@@ -5065,15 +5088,67 @@ threads_add_joinable_thread_nolock (gpointer tid)
 }
 #endif
 
+static void
+threads_add_pending_joinable_runtime_thread (gpointer thread_info)
+{
+	g_assert (thread_info);
+	MonoThreadInfo *mono_thread_info = (MonoThreadInfo*)thread_info;
+
+	if (mono_thread_info->runtime_thread) {
+		gint32 old_joinable_state = mono_atomic_load_i32 (&mono_thread_info->thread_joinable_state);
+		if (old_joinable_state == THREAD_JOINABLE_STATE_NONE) {
+			if (mono_atomic_cas_i32 (&mono_thread_info->thread_joinable_state, THREAD_JOINABLE_STATE_PENDING, THREAD_JOINABLE_STATE_NONE) == THREAD_JOINABLE_STATE_NONE) {
+				gint32 result = mono_atomic_inc_i32 (&pending_joinable_runtime_thread_count);
+				g_assert (result > 0);
+			}
+		}
+	}
+}
+
+static gboolean
+threads_wait_pending_joinable_runtime_threads (uint32_t timeout)
```

Comment:
Why is this using non coop primitives?

Reply:
See comment below, but I reused the mutex already protecting the joinable list to handle the conditional variable as well, so I inherited the non coop behavior as well.